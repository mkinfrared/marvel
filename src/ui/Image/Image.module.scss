@import "styles/colors";

.loading {
  border: none !important;
  display: inline-block !important;
  height: 0 !important;
  visibility: hidden !important;
  width: 0 !important;
}

.failed {
  align-items: center;
  background-color: $primary-main;
  display: inline-flex;
  justify-content: center;
}
