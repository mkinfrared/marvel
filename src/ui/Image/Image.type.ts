export interface ImageProps {
  alt?: string;
  loadWidth: number;
  loadHeight: number;
  loadThreshold?: number;
  src: string;
}
