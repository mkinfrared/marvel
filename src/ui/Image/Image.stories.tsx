import { Meta, Story } from "@storybook/react/types-6-0";
import React from "react";

import { Image } from "./Image";
import { ImageProps } from "./Image.type";

export default {
  title: "UI/Image",
  component: Image
} as Meta;

const Template: Story<ImageProps> = () => (
  <>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http//localhost:3001/iamge.jpg"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/city"
        loadThreshold={0.65}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/people"
        loadThreshold={1}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/transport"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/animals"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/sports/6"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/food"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/nature"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/business"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/nightlife"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/sports"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/cats"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/fashion"
        loadThreshold={0.5}
      />
    </div>
    <div>
      <Image
        loadWidth={300}
        loadHeight={300}
        src="http://lorempixel.com/300/300/technics"
        loadThreshold={0.5}
      />
    </div>
  </>
);

const Primary = Template.bind({});

export { Primary };
