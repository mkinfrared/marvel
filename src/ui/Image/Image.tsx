import React, { useEffect, useRef, useState } from "react";

import ImagePlaceholder from "ui/ImagePlaceholder";

import css from "./Image.module.scss";
import { ImageProps } from "./Image.type";

const Image = ({
  loadWidth,
  loadHeight,
  src,
  alt = "image",
  loadThreshold = 0
}: ImageProps) => {
  const [imageLoaded, setImageLoaded] = useState(false);
  const [imageFailed, setImageFailed] = useState(false);
  const imageRef = useRef<HTMLImageElement | null>(null);
  const placeholderRef = useRef<HTMLDivElement | null>(null);

  const loadImage = () => {
    if (imageRef.current) {
      imageRef.current.src = src;
    }
  };

  const config = {
    root: null,
    rootMargin: "0px",
    threshold: loadThreshold
  };

  const observer = useRef(
    new IntersectionObserver((entries, obsvr) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          loadImage();

          obsvr.unobserve(entry.target);
        }
      });
    }, config)
  );

  const handleImageLoad = () => {
    setImageLoaded(true);
  };

  const handleImageError = () => {
    setImageFailed(true);

    setImageLoaded(true);
  };

  useEffect(() => {
    const { current } = observer;

    if (placeholderRef.current) {
      current.observe(placeholderRef.current);
    }

    return () => {
      current.disconnect();
    };
  }, []);

  const className = imageLoaded ? "" : css.loading;

  return (
    <>
      {!imageFailed && (
        <img
          className={className}
          alt={alt}
          ref={imageRef}
          onLoad={handleImageLoad}
          onError={handleImageError}
        />
      )}
      {!imageLoaded && (
        <ImagePlaceholder
          width={loadWidth}
          height={loadHeight}
          forwardRef={placeholderRef}
        />
      )}
      {imageFailed && (
        <div
          className={css.failed}
          style={{ width: loadWidth, height: loadHeight }}
        >
          <p>Failed to load content</p>
        </div>
      )}
    </>
  );
};

export { Image };

export default React.memo(Image);
