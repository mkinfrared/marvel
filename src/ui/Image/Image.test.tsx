/* eslint-disable @typescript-eslint/no-empty-function,class-methods-use-this */
import { render } from "@testing-library/react";
import React from "react";

import { Image } from "./Image";

jest.mock("ui/ImagePlaceholder", () => () => "<ImagePlaceholder />");

describe("<Image />", () => {
  const Component = Image;
  const loadHeight = 42;
  const loadWidth = 47;
  const src = "marklar";
  const props = { loadHeight, loadWidth, src };

  beforeAll(() => {
    (global.IntersectionObserver as any) = class IntersectionObserver {
      constructor() {}

      disconnect() {
        return null;
      }

      observe() {
        return null;
      }

      takeRecords() {
        return null;
      }

      unobserve() {
        return null;
      }
    };
  });

  it("should be defined", () => {
    const { container } = render(<Component {...props} />);

    expect(container).toBeDefined();
  });

  it("match the snapshot", () => {
    const { container } = render(<Component {...props} />);

    expect(container).toMatchSnapshot();
  });
});
