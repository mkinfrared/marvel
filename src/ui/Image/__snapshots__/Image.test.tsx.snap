// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> match the snapshot 1`] = `
<div>
  <img
    alt="image"
    class="loading"
  />
  &lt;ImagePlaceholder /&gt;
</div>
`;
